core/explorer/extract/cli.py

```diff
@@ -0,0 +1,366 @@
+    to process to queue
+    """
+
+    # evaluate value at the attribute path using Nix
+    env = copy.deepcopy(finder_env)
+    env["TARGET_ATTRIBUTE_PATH"] = attribute_path
+    description_process = subprocess.run(
+        args=[
+            arg
+            for arg in [
+                "nix",
+                "eval",
+                "--offline" if offline else None,
+                "--extra-experimental-features",
+                "nix-command flakes",
+                "--json",
+                "--file",
+                str(pathlib.Path(__file__).parent.joinpath("describe-derivation.nix")),
+            ]
+            if arg is not None
+        ],
+        capture_output=True,
+        # write to file passed as argument (or stdout)
+        env=env,
+    )
+
+    description_str = description_process.stdout.decode()
+    if len(description_str) == 0:
+        logger.warning("Empty evaluation: %s", attribute_path)
+        logger.error(description_process.stdout)
+        return True
+
+    # add its inputs to the queue if they have not been processed
+    try:
+        description: model.Package = model.Package.parse_raw(description_str)
+    except Exception as e:
+        logger.warning(
+            "Failed to parse to model: attribute_path=%s, str=",
+            attribute_path,
+            description_str,
+        )
+        logger.error(description_process.stdout)
+        raise e from e
+
+    if description.output_path is None:
+        logger.warning("Ignore empty outputPath: %s", attribute_path)
+        return True
+
+    # write to output pipe
+    with pipe_out_lock:
+        pipe_out.write(description.json(by_alias=False))
+        pipe_out.write("\n")
+
+    visited_output_paths.add(description.output_path)
+    for build_input in description.build_inputs:
+        output_path = build_input.output_path
+        attribute_path = build_input.attribute_path
+        if output_path is not None and output_path not in queued_output_paths:
+            queue.put(attribute_path)
+            queued_output_paths.add(output_path)
+
+    # return success
+    return True
+
+
+def try_or_none(f: Callable, e_type: Type[Exception]):
+    try:
+        return f()
+    except e_type:
+        return None
+
+
+def queue_processor(
+    outfile: IO[str],
+    outfile_lock: Lock,
+    finder_process: subprocess.Popen,
+    pool: multiprocessing.pool.ThreadPool,
+    queue: multiprocessing.Queue,
+    queued_output_paths: set[str],
+    visited_output_paths: set[str],
+    finder_env: dict,
+    offline: bool,
+    logger: logging.Logger,
+):
+    """Continuously process attribute paths in the queue to the pool"""
+
+    # list of ongoing jobs
+    pool_jobs: list[multiprocessing.pool.AsyncResult] = []
+
+    # main loop of task
+    while (
+        # there is an attribute path to process
+        (
+            attribute_path := try_or_none(
+                lambda: queue.get(block=True, timeout=1),
```

Comment:
I know it's outside the scope of the review, but I'm always wary when I see blocking with timeouts. That looks like active polling and usually some kind of [notification system ](https://docs.python.org/3/library/threading.html#threading.Condition.notify) would save resources. Maybe worth considering for improvement in the future?

Reply:
Thanks. I'm not super knowledgeable about multitheaded computing, so I'll be glad to see how to improve :+1: